spynnaker/pyNN/models/neuron/synaptic_matrix_app.py

```diff
@@ -0,0 +1,719 @@
+        # this must be true in the current code, because the row length is
+        # fixed for all synaptic matrices from the same source application
+        # vertex
+        for m_edge, row_data in matrix_data:
+            size = (self.__max_row_info.undelayed_max_bytes *
+                    m_edge.pre_vertex.vertex_slice.n_atoms)
+            row_data_size = len(row_data) * BYTES_PER_WORD
+            if row_data_size != size:
+                raise Exception("Data incorrect size: {} instead of {}".format(
+                    row_data_size, size))
+            spec.write_array(row_data)
+            block_addr = self.__next_addr(block_addr, size)
+
+        return block_addr
+
+    def __write_delay_app_matrix(self, spec, block_addr, matrix_data):
+        """ Write a delay matrix for a whole incoming application vertex as one
+
+        :param DataSpecificationGenerator spec:
+            The specification to write to
+        :param int block_addr:
+            The address in the synaptic matrix region to start writing at
+        :param list(ProjectionMachineEdge, ~numpy.ndarray) matrix_data:
+            The data for each machine edge to be combined into a single matrix
+        :return: The updated block address
+        :rtype: int
+        """
+        # If there is no routing info, don't write anything
+        if self.__delay_app_key_info is None:
+            return block_addr
+
+        # If we have routing info but no synapses, write an invalid entry
+        if self.__max_row_info.delayed_max_n_synapses == 0:
+            self.__delay_index = self.__poptable.add_invalid_entry(
+                self.__delay_app_key_info.key_and_mask)
+            return block_addr
+
+        # Write a matrix for the whole application vertex
+        block_addr = self.__poptable.write_padding(spec, block_addr)
+        self.__delay_index = self.__poptable.add_application_entry(
+            block_addr, self.__max_row_info.delayed_max_words,
+            self.__delay_app_key_info.key_and_mask,
+            self.__delay_app_key_info.core_mask,
+            self.__delay_app_key_info.core_shift,
+            self.__delay_app_key_info.n_neurons)
+        self.__delay_syn_mat_offset = block_addr
+
+        # Write all the row data for each machine vertex one after the other.
+        # Implicit assumption that no machine-level row_data is ever empty;
+        # this must be true in the current code, because the row length is
+        # fixed for all synaptic matrices from the same source application
+        # vertex
+        for m_edge, row_data in matrix_data:
+            size = (self.__max_row_info.delayed_max_bytes *
+                    m_edge.pre_vertex.vertex_slice.n_atoms *
+                    self.__app_edge.n_delay_stages)
+            row_data_size = len(row_data) * BYTES_PER_WORD
+            if size != row_data_size:
+                raise Exception("Data incorrect size: {} instead of {}".format(
```

Comment:
error message at top of file is new format we seem to be going to

Reply:
esp as this is the same message

Reply:
Nope.  See above.